x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/collector/shards/ShardsCollector.java

```diff
@@ -46,24 +45,27 @@ protected Collection<MonitoringDoc> doCollect(final MonitoringDoc.Node node, fin
         if (clusterState != null) {
             RoutingTable routingTable = clusterState.routingTable();
             if (routingTable != null) {
-                List<ShardRouting> shards = routingTable.allShards();
-                if (shards != null) {
-                    final String clusterUuid = clusterUuid(clusterState);
-                    final String stateUUID = clusterState.stateUUID();
-                    final long timestamp = timestamp();
+                final String clusterUuid = clusterUuid(clusterState);
+                final String stateUUID = clusterState.stateUUID();
+                final long timestamp = timestamp();
 
-                    final String[] indices = getCollectionIndices();
-                    final boolean isAllIndices = IndexNameExpressionResolver.isAllIndices(Arrays.asList(indices));
+                final String[] indicesToMonitor = getCollectionIndices();
+                final boolean isAllIndices = IndexNameExpressionResolver.isAllIndices(Arrays.asList(indicesToMonitor));
+                final String[] indices = isAllIndices ? routingTable.indicesRouting().keySet().toArray(new String[0]) : indicesToMonitor;
 
-                    for (ShardRouting shard : shards) {
-                        if (isAllIndices || Regex.simpleMatch(indices, shard.getIndexName())) {
-                            MonitoringDoc.Node shardNode = null;
-                            if (shard.assignedToNode()) {
-                                // If the shard is assigned to a node, the shard monitoring document refers to this node
-                                shardNode = convertNode(node.getTimestamp(), clusterState.getNodes().get(shard.currentNodeId()));
-                            }
-                            results.add(new ShardMonitoringDoc(clusterUuid, timestamp, interval, shardNode, shard, stateUUID));
+                for (String index : indices) {
+                    int shardCount = 0;
+                    for (ShardRouting shard : routingTable.allShards(index)) {
+                        if (shard.primary()) {
+                            shardCount = 0;
```

Comment:
This looping logic worries me. 

Based on the routing table code, there's nothing to suggest that shards returned from it will always be ordered such that the primary is returned before the replicas. Instead, it looks like the order of the shards is based entirely on how the table is created, and that is dictated by code that lives far away from here.  

If the ordering isn't correct, then this logic could lead to id collisions for shards. If the returned shards are ordered `replica1, replica1, primary1, replica1` then you'll only get three distinct id's for those shard copies. The shard routings in the second and fourth slots will have the same counter value when the primary in slot 3 resets it. Furthermore, the next shard in the list will have its counters offset unless it has the primary sorted first.

Since we'll need to get the index routing table above to make sure it's not null before using it, let's build further on that. Let's get the number of shards from the index routing table with `routingTable.size()`. Then let's do a for loop bound by the shard count, asking for the shard routing table for each shard number with `routingTable.shard(idx)`. After that, we can ask each shard routing table for its primary shard and iterate over any replicas it has with safer counting logic. What do you think?

Reply:
This makes a lot of sense to me.

I just wonder about one thing, the method `IndexShardRoutingTable.replicaShards` returns a list of all replica shards but isn't the order of those replica shards also up to whatever code instantiates the `IndexShardRoutingTable`, so they might also be out of order? 

That might be a much smaller problem than the skipping index code I had before with the primaries marking the reset point. I'm not sure if it would be a big problem if the order of the replica shards changed?